docs/ops/matrix/MatMul_1.md

```diff
@@ -12,29 +12,43 @@
 
 Before matrix multiplication, there is an implicit shape alignment for input arguments. It consists of the following steps:
 
-1. If rank of an input less than 2 it is unsqueezed to 2D tensor by adding axes with size 1 to the left of the shape. For example, if input has shape `[S]` it will be reshaped to `[1, S]`. It is applied for each input independently.
+1. One-dimensional tensors unsqueezing is applied for each input independently. The axes inserted in this step are not included in the output shape.
+    * If rank of the **first** input is less than 2, it is unsqueezed to 2D tensor by adding axes with size 1 to the **left** of the shape. For example `[S]` will be reshaped to `[1, S]`.
+    * If rank of the **second** input is less than 2, it is unsqueezed to 2D tensor by adding axes with size 1 to the **right** of the shape. For example `[S]` will be reshaped to `[S, 1]`.
 
 2. Applied transpositions specified by optional `transpose_a` and `transpose_b` attributes.
```

Comment:
Here are the steps for current [MatMul reference implementation](https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/reference/include/ngraph/runtime/reference/matmul.hpp#L65-L71):

> // Steps to compute matmul:
> // 1) Check inputs and perform transpose on arg if applicable
> // 2) If ranks of both args are 2D and below (no batch dim),
> // perform dot and return result; otherwise, continue next
> // 3) Check if auto broadcast is needed on args or transposed args,
> // and perform broadcast if applicable
> // 4) Perform dot on the args or updated args and return result

So the `transpose_a` and `transpose_b` params are ignored for 1D.
I just want to ensure that we don't want to follow this logic anymore and apply the transpose after 1D to 2D unsqueezing.

Reply:
Yes, according to our recent conversations, we shouldn't apply transposes to 1D tensors to avoid introducing an new MatMul version. @lazarevevgeny  and @jane-intel please approve. @mitruska please change PR accordingly, it still says that we should apply transposes after unsqueezing 1D tensors.

Reply:
Apart from MatMul reference implementation,
here we have a conflict between ngraph MatMul shape inference, transformation to Gemm and ModelOptimizer.

These "implicit shape alignment steps" describe what **Model Optimizer** do.
https://github.com/openvinotoolkit/openvino/blob/0ef1fb8a097e91145c694bef85dc50c4341e92a9/model-optimizer/extensions/ops/MatMul.py#L26-L140

The **transformation to Gemm** actually performs current spec step number 1 and 3 at once.

https://github.com/openvinotoolkit/openvino/blob/0ef1fb8a097e91145c694bef85dc50c4341e92a9/inference-engine/src/legacy_api/src/transformations/convert_opset1_to_legacy/convert_matmul_to_fc_or_gemm.cpp#L166-L235

It expands smaller tensor rank to bigger tensor rank and if B input is 1D, additional 1 dim is inserted at the end so [1..., S, 1]. This is also why 1D x 1D is not supported by this transformation because handling equal ranks is missed. Transpose is applied later, not in this transformation.
For output from 1D input cases multiplying it also squeezes this additional axis by inserting Reshape to original ngraph output (numpy/onnx) at the end.

Legacy **ngraph MatMul fused op** decomposition [(matmul builder)](https://github.com/openvinotoolkit/openvino/blob/28eed7708e76dfa605f3078935244527d783d0ac/ngraph/core/builder/src/builder/matmul_factory.cpp#L74-L89) and validates 1D always as A[1, S] or B[S, 1] and corresponding transpose_a/b as `false`.
**So 1D input shapes that would be compatible only after 1D transposition are rejected anyway at first in MatMul op input validation.**

As I checked, it's about few lines of code to enable 1D transposition by not ignoring 1D transpose attributes during input validation and output shape inference in MatMul `validate_and_infer_types()`.
We can consider it as a future fix/change.

In my opinion to describe current ngraph MatMul 1D transpose situation we can only say:
"Transposition of 1D inputs by `transpose_a` or `transpose_b` is not supported".
But it's not true about ModelOptimizer or the transformation so I suggest to left this PR with only 1D unsqeezing and output info update (current changes in PR).

Reply:
Does "So 1D input shapes that would be compatible only after 1D transposition are rejected anyway at first in MatMul op input validation" means exactly that we don't apply transpositions? I'm a bit lost. Currently we are talking only about specification. So you want to apply more strong statement that they are not supported, instead of "not applied", right?

Reply:
I'm just saying that what is true about one part of the project is not necessary true about second and it's hard to find generic description that is suitable for each part. 
Ngraph MatMul input validation is not aware of 1D transform attributes, but MO and the transposition to Gemm are able to handle that.

Reply:
We had a meeting with @mitruska and found a case when potential change in the current behavior leads to a backward incompatibility. The potential change is to start applying transposes to 1D input tensors. We cannot do that and still need to follow exactly the same algorithm as original MatMul validation function implemented. It was decided that @mitruska was going to change the spec accordingly.